Cody: Add docs on generate index (#52206)
sourcegraph/sourcegraph: doc/cody/explanations/indexing.md

# Generate Index to Enable Codebase-Aware Answers

> This docs provides instructions on how to generate an index that enables codebase-aware answers for Cody. Codebase-aware answers leverage code graph context to enhance Cody's understanding of code from selected codebases. 

Reviewer: ```suggestion
These docs provide instructions on how to generate an index that enables codebase-aware answers for Cody. Codebase-aware answers leverage code graph context to enhance Cody's understanding of code from selected codebases. 
```

By following the steps outlined in this guide, you can configure the necessary prerequisites and enable Cody to provide more accurate and contextually relevant answers to your coding questions based on the codebase you are working in.

## Generate Index

You can enhance Cody's understanding of existing code by generating index for your codebases to enable [code graph context](https://docs.sourcegraph.com/cody/explanations/code_graph_context).

### Sourcegraph Enterprise

> To generate an index for your codebase and enable codebase-aware answers for Cody, your Site Admins must complete the following:

Reviewer: ```suggestion
To generate an index for your codebase and enable codebase-aware answers for Cody, your site admin must complete the following:
```

- [Configure Code Graph Context](https://docs.sourcegraph.com/cody/explanations/code_graph_context) for your Sourcegraph instance
- [Enable Cody for your Sourcegraph instance](../enabling_cody_enterprise.md#step-1-enable-cody-on-your-sourcegraph-instance)
- [Enable Cody for your Sourcegraph account](../enabling_cody_enterprise.md#turning-cody-off)

### Sourcegraph.com

> Code graph context is available instantly for public repositories on Sourcegraph.com that are already embedded. Please refer to the [list of repositories with embeddings](https://docs.sourcegraph.com/cody/embedded-repos) for the latest information.

Reviewer: How up-to-date is this list?

Reviewer: You seem to be using "code graph context" and "embeddings" interchangeably (unless I'm misinterpreting and you intend them to be different). I think we should prefer "code graph context", because embeddings are more of  an implementation detail.

Author: Update to remove any mentions of embedded except the link to the repo list and channel name

Code graph context is not available for codebases that are not included in the list. 

However, you can make requests and ask for assistance in the [Sourcegraph Discord](https://discord.gg/8wJF5EdAyA) under the `#cody-embeddings` channel.

> NOTE: Sourcegraph.com does not support connections to private repositories

## Enable Codebase-Aware Answers

> To enable codebase-aware answers for the Cody extension, you must set the `Cody: Codebase` (`cody.codebase`) configuration option in VS Code to the repository name on your Sourcegraph instance. By doing so, Cody will provide more accurate and relevant answers to your coding questions, referencing to the context of the codebase you are currently working in.

Reviewer: Is this still true? Do we not infer the codebase ID from the clone URL of the git repository now? When does Cody use `cody.codebase` (i.e., does setting this override the inferred codebase ID)?

By following the steps outlined in this guide, you can configure the necessary prerequisites and enable Cody to provide more accurate and contextually relevant answers to your coding questions based on the codebase you are working in.

## Generate Index

You can enhance Cody's understanding of existing code by generating index for your codebases to enable [code graph context](https://docs.sourcegraph.com/cody/explanations/code_graph_context).

### Sourcegraph Enterprise

To generate an index for your codebase and enable codebase-aware answers for Cody, your site admin must complete the following:

- [Configure Code Graph Context](https://docs.sourcegraph.com/cody/explanations/code_graph_context) for your Sourcegraph instance
- [Enable Cody for your Sourcegraph instance](../enabling_cody_enterprise.md#step-1-enable-cody-on-your-sourcegraph-instance)
- [Enable Cody for your Sourcegraph account](../enabling_cody_enterprise.md#turning-cody-off)

### Sourcegraph.com

>  See the current [list of repositories with code graph context available](https://docs.sourcegraph.com/cody/embedded-repos), and request any that you'd like to add by pinging a Sourcegraph team member in the [Sourcegraph Discord](https://discord.gg/8wJF5EdAyA) under the `#cody-embeddings` channel.

Author: this list was updated by @kalanchan today in https://github.com/sourcegraph/sourcegraph/pull/52568. He will also work on getting the list to update automatically.

> NOTE: Sourcegraph.com does not support connections to private repositories

## Enable Codebase-Aware Answers

When connected to the correct codebase, Cody provides accurate and relevant answers to your coding questions, taking into account the context of the codebase you are currently working in.

Cody attempts to connect to the appropriate codebase automatically using Git, eliminating the need for manual configuration of the `Cody: Codebase` (`cody.codebase`) option. 

You only need to configure the `Cody: Codebase` (`cody.codebase`) setting if Cody displays `NOT INDEXED` below the chat window.

### Extension Settings

Cody tries to connect to the correct codebase using Git at start-up, so manual configuration of `cody.codebase` is unnecessary.

> To enable Cody to connect to the codebase using Git, ensure that the `Cody: Codebase` (`cody.codebase`) configuration field is empty.

Reviewer 2: ```suggestion
To enable Cody to set the codebase config using `git remote get-url origin`, ensure that the `Cody: Codebase` (`cody.codebase`) configuration field is unset in your VS Code User and Workspace settings, and remote workspace and folder-level settings if applicable.
```

If this attempt fails and you see `NOT INDEXED` below your Cody chatbox, you can manually specify the correct codebase for Cody using the `Cody: Codebase` (`cody.codebase`) configuration option.

#### Manual Configuration

Follow these steps to manually configure the `codebase` setting for Cody via the [Extension Settings](https://code.visualstudio.com/docs/getstarted/settings#_extension-settings) in VS Code:

1. Open the VS Code workspace settings by clicking: 
   - Mac: `Code` > `Settings` > `Settings`
   - Windows & Linux: `File` > `Preferences (Settings)`

> 2. Enter `Cody: Codebase` in the search bar

Reviewer 2: ```suggestion
2. Switch to the Workspace settings tab
3. Enter `Cody: Codebase` in the search bar
```

> 3. Enter the repository name as listed on your Sourcegraph instance in the `Cody: Codebase` field

Reviewer 2: ```suggestion
4. Enter the repository name as listed on your Sourcegraph instance in the `Cody: Codebase` field
```
